orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,503 @@
+		// get a puller
+
+		go func(ec EndpointCriteria, env *common.Envelope) {
+			defer waitGrp.Done()
+			block, err := p.pullAttestationBlock(ec, seq, env)
+			if err != nil {
+				p.Logger.Errorf("errors in pulling attestations: %v", err)
+			}
+			bpsInfo <- &bps{block, ec}
+		}(ec, env)
+	}
+	waitGrp.Wait()
+	//close channel
+	close(bpsInfo)
+
+	attestations := []*orderer.BlockAttestation{}
+	for b := range bpsInfo {
+		// verify the blocks or attestations returned here and compare with the
+		// previously pulled block
+		attestations = append(attestations, b.block)
+	}
+	return attestations
+}
+
+//FetcherConfig stores the configuration parameters needed to create a BlockFetcher
+type FetcherConfig struct {
+	Channel      string
+	Signer       identity.SignerSerializer
+	TLSCert      []byte
+	Dialer       Dialer
+	Endpoints    []EndpointCriteria
+	FetchTimeout time.Duration
+}
+
+type TimeFunc func() time.Time
+
+//BlockFetcher can be used to fetch blocks from orderers in a byzantine fault tolerant way.
+type BlockFetcher struct {
+	MaxPullBlockRetries      uint64
+	AttestationSourceFactory func(c FetcherConfig) AttestationSource
+	BlockSourceFactory       func(c FetcherConfig) BlockSource
+	Config                   FetcherConfig
+	suspects                 suspectSet // a set of bft suspected nodes.
+	Logger                   *flogging.FabricLogger
+	currentBlockSource       BlockSource
+	currentEndpoint          EndpointCriteria
+	currentSeq               uint64
+	ShuffleTimeout           time.Duration
+	LastShuffledAt           time.Time
+	MaxByzantineNodes        uint64
+	ConfirmByzantineBehavior func(attestations []*orderer.BlockAttestation) bool
+	ShuffleTimeoutThrehold   int64 // expressed as %age of FetchTimeOut
+	TimeNow                  TimeFunc
+}
+
+func (bf *BlockFetcher) getBlockSource(op BlockSourceOps) BlockSource {
+
+	switch op {
+	case CURRENTSOURCE:
+		// set the first source or shuffle the current source
+		emptyEC := EndpointCriteria{}
+		if bf.isShuffleTime() || bf.currentEndpoint.Endpoint == emptyEC.Endpoint {
+			bf.shuffleEndpoint()
+		}
+	case SHUFFLESOURCE:
+		bf.shuffleEndpoint()
+	case SUSPECTSOURCE:
+		blockwitheld, err := bf.isBlockWitheld(bf.currentSeq)
+		if blockwitheld || err != nil {
+			bf.shuffleEndpoint()
+		}
+	}
+
+	return bf.currentBlockSource
+}
+
+func (bf *BlockFetcher) setBlockSource(ec EndpointCriteria) {
+
+	// close old blocksource
+	if bf.currentBlockSource != nil {
+		bf.currentBlockSource.Close()
+	}
+	// create a local config
+	config := bf.Config
+	// fill only current endpoint criteria
+	config.Endpoints = []EndpointCriteria{ec}
+	bf.currentBlockSource = bf.BlockSourceFactory(config)
+	bf.currentBlockSource.UpdateEndpoints([]EndpointCriteria{ec})
+}
+
+func (bf *BlockFetcher) shuffleEndpoint() error {
+	// When this function is called for the first time
+	// set the currentendpoint
+	emptyEC := EndpointCriteria{}
+
+	if bf.currentEndpoint.Endpoint != emptyEC.Endpoint {
```

Comment:
why not do something like the following?

1. Get all endpoints you have in total, let's call this list `L`
2. Remove all suspects from `L`
3. Remove current endpoint from `L`
4. Return endpoint in location L[rand(len(L)]